File: frontend/src/lib/api/reviewQueue.ts
/**
 * Review Queue API Client
 *
 * Phase 162: Semantic Control Foundation — Plan 05.
 * Human-in-the-loop review interface for semantic classification decisions.
 */

import { fetchApi } from "./client";

export interface ReviewQueueEntry {
  id: string;
  site_id: string;
  equipment_id: string;
  point_id: string;
  classification_id: string;
  semantic_tags: string[];
  confidence_score: number;
  confidence_level: string; // HIGH, MEDIUM, LOW
  safety_class: string; // LOW, MEDIUM, HIGH
  automation_tier: string; // observe_only, supervised, automatic
  validation_passed: boolean;
  validation_errors: unknown[];
  completeness_score: number | null;
  status: string; // pending, approved, rejected, overridden
  priority: number;
  classified_by: string;
  classified_at: string;
  reviewed_by: string | null;
  reviewed_at: string | null;
  review_notes: string | null;
  decision_reason: string | null;
  override_tags: string[] | null;
  override_justification: string | null;
}

export interface ReviewQueueStats {
  total_pending: number;
  by_safety_class: Record<string, number>;
  by_confidence_level: Record<string, number>;
  avg_age_hours: number;
  high_priority_count: number;
}

export interface ReviewDecisionResponse {
  entry_id: string;
  success: boolean;
  message: string;
}

export interface BulkDecisionResponse {
  approved_count: number;
  message: string;
}

export interface PendingFilters {
  site_id: string;
  safety_class?: string;
  equipment_id?: string;
  confidence_threshold?: number;
  limit?: number;
}

export const reviewQueueApi = {
  async getPending(filters: PendingFilters): Promise<ReviewQueueEntry[]> {
    const params = new URLSearchParams({ site_id: filters.site_id });
    if (filters.safety_class) params.set("safety_class", filters.safety_class);
    if (filters.equipment_id) params.set("equipment_id", filters.equipment_id);
    if (filters.confidence_threshold !== undefined) {
      params.set("confidence_threshold", String(filters.confidence_threshold));
    }
    if (filters.limit !== undefined) {
      params.set("limit", String(filters.limit));
    }
    return fetchApi<ReviewQueueEntry[]>(`/api/review-queue/pending?${params.toString()}`);
  },

  async getStats(site_id: string): Promise<ReviewQueueStats> {
    return fetchApi<ReviewQueueStats>(`/api/review-queue/stats?site_id=${encodeURIComponent(site_id)}`);
  },

  async approve(entry_id: string, review_notes: string): Promise<ReviewDecisionResponse> {
    return fetchApi<ReviewDecisionResponse>(`/api/review-queue/${encodeURIComponent(entry_id)}/approve`, {
      method: "POST",
      body: JSON.stringify({ review_notes }),
    });
  },

  async reject(entry_id: string, reason: string, review_notes: string): Promise<ReviewDecisionResponse> {
    return fetchApi<ReviewDecisionResponse>(`/api/review-queue/${encodeURIComponent(entry_id)}/reject`, {
      method: "POST",
      body: JSON.stringify({ reason, review_notes }),
    });
  },

  async override(
    entry_id: string,
    correct_tags: string[],
    justification: string
  ): Promise<ReviewDecisionResponse> {
    return fetchApi<ReviewDecisionResponse>(`/api/review-queue/${encodeURIComponent(entry_id)}/override`, {
      method: "POST",
      body: JSON.stringify({ correct_tags, justification }),
    });
  },

  async bulkApprove(entry_ids: string[]): Promise<BulkDecisionResponse> {
    return fetchApi<BulkDecisionResponse>("/api/review-queue/bulk-approve", {
      method: "POST",
      body: JSON.stringify({ entry_ids }),
    });
  },

  async getHistory(entry_id: string): Promise<unknown[]> {
    return fetchApi<unknown[]>(`/api/review-queue/${encodeURIComponent(entry_id)}/history`);
  },
};

File: backend/app/models/review_queue.py
"""Pydantic models for review queue entries and decisions.

Phase 162: Semantic Control Foundation — Plan 05.
Human-in-the-loop review interface for semantic classification decisions.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ReviewQueueEntry(BaseModel):
    """A classification queued for human review."""

    id: str | None = None
    site_id: str
    equipment_id: str
    point_id: str
    classification_id: str

    # Classification details
    semantic_tags: list[str] = Field(default_factory=list)
    confidence_score: float = Field(ge=0.0, le=1.0)
    confidence_level: str  # HIGH, MEDIUM, LOW
    safety_class: str  # LOW, MEDIUM, HIGH
    automation_tier: str  # observe_only, supervised, automatic

    # Validation results
    validation_passed: bool = False
    validation_errors: list[Any] = Field(default_factory=list)
    completeness_score: float | None = None

    # Review metadata
    status: str = "pending"  # pending, approved, rejected, overridden
    priority: int = 100  # Lower = higher priority

    # Audit trail
    classified_by: str
    classified_at: datetime = Field(default_factory=datetime.utcnow)

    # Review decision
    reviewed_by: str | None = None
    reviewed_at: datetime | None = None
    review_notes: str | None = None
    decision_reason: str | None = None

    # Override data
    override_tags: list[str] | None = None
    override_confidence: float | None = None
    override_justification: str | None = None

    # Trust reset metadata (PLAN-162B)
    reset_reason: str | None = None

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class ReviewDecision(BaseModel):
    """A decision made on a review queue entry."""

    id: str | None = None
    review_queue_id: str
    decision_type: str  # approve, reject, override
    decision_reason: str | None = None
    reviewed_by: str
    reviewed_at: datetime = Field(default_factory=datetime.utcnow)
    review_notes: str | None = None
    metadata: dict | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class ReviewQueueStats(BaseModel):
    """Statistics about the review queue."""

    total_pending: int = 0
    by_safety_class: dict[str, int] = Field(default_factory=dict)
    by_confidence_level: dict[str, int] = Field(default_factory=dict)
    avg_age_hours: float = 0.0
    high_priority_count: int = 0  # priority <= 50
